### PR TITLE
Split and move around exceptions, make MigrationException an interface

### LINKED ISSUE #621: Split up monster MigrationException in to smaller exceptions
This same thing is being done in the DBAL and the ORM

- https://github.com/doctrine/dbal/pull/3131
- https://github.com/doctrine/doctrine2/pull/6743
- https://github.com/doctrine/doctrine2/pull/7210

### PR DESCRIPTION
**Pull request by Majkl578**
* root MigrationException is now an interface
* factory methods from MigrationException extracted to specialized exceptions
* exceptions from root moved to Exception sub-namespace
* all exceptions are now final
* all exceptions now implement MigrationException (directly or indirectly) and extending SPL exceptions

This follows concept introduced in ORM (https://github.com/doctrine/doctrine2/pull/6743 + https://github.com/doctrine/doctrine2/pull/7210) and naming follows pattern accepted in Doctrine CS.

No polyfill is currently provided in this PR, it should be introduced in 1.x afterwards.

closes #621 

cc @greg0ire

### CODE REVIEW

File: lib/Doctrine/Migrations/Configuration/Exception/FileAlreadyLoaded.php [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+
+final class FileAlreadyLoaded extends LogicException implements ConfigurationException
+{
+    public static function new() : self
+    {
+        return new self(
+            sprintf(
+                'Migrations configuration file already loaded'
+            ),
+            8
```
[the remark this answers is not in the document]
Comment (greg0ire) [reply lib/Doctrine/Migrations/Configuration/Exception/FileAlreadyLoaded.php]: Ah I see, they were there before, let's keep them then, I suppose.

File: lib/Doctrine/Migrations/Configuration/Exception/FileAlreadyLoaded.php [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+
+use LogicException;
+use function sprintf;
+
+final class FileAlreadyLoaded extends LogicException implements ConfigurationException
+{
+    public static function new() : self
+    {
+        return new self(
+            sprintf(
```
Comment (lcobucci) [inline lib/Doctrine/Migrations/Configuration/Exception/FileAlreadyLoaded.php]: `sprintf()` here is not really needed

File: lib/Doctrine/Migrations/Configuration/Exception/FileAlreadyLoaded.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (Majkl578) [reply lib/Doctrine/Migrations/Configuration/Exception/FileAlreadyLoaded.php]: You will have to ask someone else for that. :P https://github.com/doctrine/migrations/commit/5d6fbf8089db10f6df5ce607400b6731183b44d0
Just kept those for compatibility.

[the remark this answers is not in the document]
Comment (Majkl578) [reply lib/Doctrine/Migrations/Configuration/Exception/FileAlreadyLoaded.php]: OF course, with this new exception layout it's obsolete.

**Comment by Majkl578** [comment]
I further split Doctrine\Migrations\Configuration\Exception\NotValid to specific ones and removed it.

**Comment by jwage** [comment]
Ready to merge?

**Comment by Majkl578** [comment]
Still wanted @alcaeus to have a look. 😊 But if you need it in for some other work, could be discussed after merge.

**Comment by Majkl578** [comment]
Hmm, I [alegedly decreased coverage](https://scrutinizer-ci.com/g/doctrine/migrations/inspections/f834cbae-d46d-4a83-a978-6f4dcf08c5d7/code-structure/test-coverage) by 1% just by moving around the methods and it's now blocking merge. :thinking:

**Pull request merged**